libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/DefaultTracer.java

```diff
@@ -27,31 +27,37 @@ class DefaultTracer implements Tracer {
     static final String THREAD_NAME = "th_name";
 
     private final TracingTelemetry tracingTelemetry;
-    private final TracerContextStorage<String, Span> tracerContextStorage;
+    private final TracerContextStorage<String, Span> spanTracerContextStorage;
+    private final ThreadLocal<SpanScope> spanScopeThreadLocal = new ThreadLocal<>();
```

Comment:
Please move it off to `DefaultSpanScope` (the scope should be keeping its thread locals, not tracer)

Reply:
I kept it outside only for one case where on the synchronous call if somebody uses SpanScope instead of ScopedSpan then Span.endSpan() will not update the TracerContextStorage. Span.endSpan() will never update the TracerContextStorage but for async call it will always be a starting point so it doesn't matter but in the synchronous call span.endSpan() should set the parent as a current span.

```
            final Span httpRequestSpan = tracer.startSpan(SpanBuilder.from(httpRequest));
            // Attaches the span to current thread (scope)
            final (final SpanScope scope = tracer.spanInScope(httpRequestSpan)) {
                doWork();
                httpRequestSpan.endSpan(); **// This will not update the TracerContextStorage with the parentSpan.**
            } 
```

Otherwise we need to be very explicit that in the synchronous path we should always use the ScopedSpan.

Reply:
Even in the Async case also (ex below) once the SpanScope is closed we need to reset the TracerContextStorage also for that thread where span is created.

```
// Creates a Span that is not associated with the thread (it is not attached to anything)
            final Span httpRequestSpan = tracer.startSpan(SpanBuilder.from(httpRequest));
            // Attaches the span to current thread (scope)
            final (final SpanScope scope = tracer.spanInScope(httpRequestSpan)) {
                HttpChannel traceableHttpChannel = new TraceableHttpChannel(httpChannel, httpRequestSpan);
                handleIncomingRequest(httpRequest, traceableHttpChannel, httpRequest.getInboundException());
            } 
```

Reply:
> Otherwise we need to be very explicit that in the synchronous path we should always use the ScopedSpan.

The sync flow either should use `ScopedSpan` or Span + Scope:

```
final Span span = tracer.startSpan(...);
final (final SpanScope scope = tracer.spanInScope(httpRequestSpan)) { 
    doWork();
} finally {
   span.end();
}
```

Those should lead to the same results. Now, why span.end()  doesn't update the TracerContextStorage?

Reply:
TracerContextStorage is as of now managed by the tracer and Span::endSpan operation happens on the Span directly. We need to pass that closable to span in that case to achieve the same behavior. Let me check the feasibility

Reply:
As part of tracingTelemetry::createSpan, we can pass on onEndConsumer which can help with updating the TracerContextStorage. Let me try that.

Reply:
We need to be careful here: if Span picks the parent from the `tracerContextStorage` on start, only than we should attempt to restore it on `end`. Also, before doing that,  the span should check it is still the current one in `tracerContextStorage`, and it not - do nothing.

Reply:
Parent can be attached to a span at start in 3 ways,
1. User provided explicitly.
2. From tracerContextStorage
3. From header
4. Span from context propagation have some stale parent from the previous thread.

I think except point 4, we have a valid parent and we should set it. Any case apart form 1,2 and 3 we should set null.

Reply:
> Any case apart form 1,2 and 3 we should set null.

Probably we should throw an exception here instead, this violates the context propagation and storage, we should better catch it.

Reply:
4th one is a valid scenario but ideally we shouldn't update the context with its's parent and this is always be a first span so we can either update with null or leave it as it is as stashContext will take care of it. Other than that we can throw the exception. 

 What do you suggest, How should we catch the exception - Set as an error to the span or throw all the way to caller

Reply:
Sorry, the exception was a bad idea by all means (crossed it out), because in many cases span.end() would be called from different thread. 

>  leave it as it is as stashContext will take care of it. 

that's I think for now it the best option

Reply:
@reta, Updated the code. Please take a look.